cloud-aws-common/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/common/resource/volume/AwsVolumeResourceBuilder.java

```diff
@@ -367,21 +370,130 @@ public CloudResource delete(AwsContext context, AuthenticatedContext auth, Cloud
 
         boolean anyDeleted = cloudResourceStatuses.stream().map(CloudResourceStatus::getStatus).anyMatch(DELETED::equals);
         AmazonEc2Client client = getAmazonEC2Client(auth);
-        if (!volumeSetAttributes.getDeleteOnTermination() && !anyDeleted) {
-            LOGGER.debug("Volumes will be preserved.");
-            resource.setStatus(CommonStatus.DETACHED);
-            volumeSetAttributes.setDeleteOnTermination(Boolean.TRUE);
-            turnOffDeleteOnterminationOnAttachedVolumes(resource, cloudResourceStatuses, client);
-            resource.putParameter(CloudResource.ATTRIBUTES, volumeSetAttributes);
-            resourceNotifier.notifyUpdate(resource, auth.getCloudContext());
-            throw new PreserveResourceException("Resource will be preserved for later reattachment.");
+        if (!volumeSetAttributes.getDeleteOnTermination()) {
+            if (anyDeleted) {
+                LOGGER.debug(String.format("Volumes should be preserved for instance (%s), but some of them are already terminated. " +
+                        "Trying to recover/recreate them.", resource.getInstanceId()));
+                List<Volume> recoveredVolumes = tryRecoverVolumeSet(context, auth, resource, volumeSetAttributes);
+                LOGGER.debug(String.format("Recovered/recreated volumes: %s", recoveredVolumes));
+                if (recoveredVolumes != null && !recoveredVolumes.isEmpty()) {
+                    volumeSetAttributes.setVolumes(recoveredVolumes);
+                    setVolumeDetachParameters(auth, resource, volumeSetAttributes, cloudResourceStatuses, client);
+                }
```

Comment:
for safety reason shouldn't be better to move setVolumeDetachParameters(auth, resource, volumeSetAttributes, cloudResourceStatuses, client);  out of the above if branch?